File: mediapipe/tasks/java/com/google/mediapipe/tasks/retrieval/semanticretriever/SemanticRetriever.java
/* Copyright 2026 The MediaPipe Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package com.google.mediapipe.tasks.retrieval.semanticretriever;

import android.content.Context;
import android.net.Uri;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.mediapipe.tasks.components.containers.AudioData;
import com.google.mediapipe.tasks.components.containers.Embedding;
import com.google.mediapipe.tasks.components.utils.CosineSimilarity;
import com.google.mediapipe.tasks.core.AudioPart;
import com.google.mediapipe.tasks.core.EmbeddingProvider;
import com.google.mediapipe.tasks.core.ImagePart;
import com.google.mediapipe.tasks.core.Part;
import com.google.mediapipe.tasks.core.TextPart;
import com.google.mediapipe.tasks.core.logging.TasksStatsLogger;
import com.google.mediapipe.tasks.core.logging.TasksStatsLoggerFactory;
import com.google.mediapipe.tasks.retrieval.chunking.DefaultTextChunker;
import com.google.mediapipe.tasks.retrieval.chunking.TextChunker;
import com.google.mediapipe.tasks.retrieval.model.RetrievalRecord;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.atomic.AtomicLong;

/** MediaPipe SemanticRetriever API that manages vector embedding and retrieval. */
@SuppressWarnings({"IfChainToSwitch", "PatternMatchingInstanceof"})
public final class SemanticRetriever implements AutoCloseable {
  private static final int DEFAULT_CHUNK_SIZE = 512;
  private static final int DEFAULT_CHUNK_OVERLAP = 100;

  private final Context context;
  private final SemanticRetrieverComponents components;
  private final TasksStatsLogger statsLogger;
  private final AtomicLong syntheticTimestamp = new AtomicLong(0);
  private final TextChunker textChunker;

  /**
   * Represents a prepared retrieval record and its parent/child relationships prior to vector
   * embedding generation and database insertion.
   */
  private static final class DatabaseRecord {
    final String id;
    final Part part;
    final String parentId;
    final List<String> childIds;

    DatabaseRecord(String id, Part part, String parentId, List<String> childIds) {
      this.id = id;
      this.part = part;
      this.parentId = parentId;
      this.childIds = childIds;
    }
  }

  private SemanticRetriever(Context context, SemanticRetrieverComponents components) {
    this.context = context;
    this.components = components;
    this.statsLogger =
        TasksStatsLoggerFactory.create(context, "SemanticRetriever", /* taskRunningModeStr= */ "");
    this.statsLogger.logSessionStart();
    this.textChunker =
        components.textChunker() != null
            ? components.textChunker()
            : new DefaultTextChunker(
                DEFAULT_CHUNK_SIZE,
                DEFAULT_CHUNK_OVERLAP,
                DefaultTextChunker.ChunkingMode.CHARACTER);
  }

  /**
   * Creates a {@link SemanticRetriever} from {@link SemanticRetrieverComponents}.
   *
   * @param context the Android context.
   * @param components the SemanticRetriever components.
   * @return a SemanticRetriever instance.
   */
  public static SemanticRetriever createFromComponents(
      Context context, SemanticRetrieverComponents components) {
    return new SemanticRetriever(context, components);
  }

  /**
   * Inserts a document into the vector store.
   *
   * @param recordId the id of the record to insert.
   * @param text the text of the record to insert.
   */
  public void insertDocument(String recordId, String text) {
    insertDocument(recordId, text, ImmutableMap.of());
  }

  /**
   * Inserts a document into the vector store with metadata.
   *
   * @param recordId the id of the record to insert.
   * @param text the text of the record to insert.
   * @param metadata the metadata of the record to insert.
   */
  public void insertDocument(String recordId, String text, Map<String, String> metadata) {
    insertContent(recordId, ImmutableList.of(new TextPart(text)), metadata);
  }

  /** Inserts an image into the vector store using a Uri. */
  public void insertImage(String recordId, Uri uri) {
    insertImage(recordId, uri, ImmutableMap.of());
  }

  /** Inserts an image into the vector store using a Uri and metadata. */
  public void insertImage(String recordId, Uri uri, Map<String, String> metadata) {
    insertContent(recordId, ImmutableList.of(new ImagePart(uri)), metadata);
  }

  /** Inserts audio into the vector store using a Uri. */
  public void insertAudio(String recordId, Uri uri) {
    insertAudio(recordId, uri, ImmutableMap.of());
  }

  /** Inserts audio into the vector store using a Uri and metadata. */
  public void insertAudio(String recordId, Uri uri, Map<String, String> metadata) {
    insertContent(recordId, ImmutableList.of(new AudioPart(uri)), metadata);
  }

  /**
   * Inserts content into the vector store.
   *
   * @param recordId the id of the record to insert.
   * @param parts the parts to insert.
   */
  public void insertContent(String recordId, List<Part> parts) {
    insertContent(recordId, parts, ImmutableMap.of());
  }

  /**
   * Inserts content into the vector store with metadata.
   *
   * @param recordId the id of the record to insert.
   * @param parts the parts to insert.
   * @param metadata the metadata of the record to insert.
   */
  public void insertContent(String recordId, List<Part> parts, Map<String, String> metadata) {
    // Delete any existing parent record and its associated child chunk records to prevent
    // orphaning.
    delete(ImmutableList.of(recordId));

    List<DatabaseRecord> databaseRecords = new ArrayList<>();
    int childCount = 0;

    // Phase 1: Iterate through parts and prepare database records
    for (Part part : parts) {
      if (part instanceof TextPart) {
        TextPart textPart = (TextPart) part;
        List<String> chunks = textChunker.chunk(textPart.getText());
        if (chunks.size() > 1) {
          List<String> childIds = new ArrayList<>();
          for (String chunkText : chunks) {
            String childId = recordId + "_chunk_" + childCount++;
            childIds.add(childId);
            databaseRecords.add(
                new DatabaseRecord(childId, new TextPart(chunkText), recordId, null));
          }
          databaseRecords.add(new DatabaseRecord(recordId, textPart, null, childIds));
          continue;
        }
      }
      databaseRecords.add(new DatabaseRecord(recordId, part, null, null));
    }

    // Phase 2: Generate embeddings and insert into vector store
    for (DatabaseRecord record : databaseRecords) {
      float[] embedding = embedContent(ImmutableList.of(record.part));
      persistPart(record.id, record.part, embedding, metadata, record.parentId, record.childIds);
    }
  }

  private void persistPart(
      String recordId,
      Part part,
      float[] embeddings,
      Map<String, String> metadata,
      String parentId,
      List<String> childIds) {
    ImmutableList.Builder<RetrievalRecord> records = ImmutableList.builder();
    records.add(
        new RetrievalRecord(
            recordId, ImmutableList.of(part), embeddings, metadata, parentId, childIds));
    components.vectorStore().upsert(records.build());
  }

  private float[] embedContent(List<Part> content) {
    List<Object> resolvedContent = new ArrayList<>();
    for (Part part : content) {
      if (part instanceof ImagePart) {
        ImagePart imagePart = (ImagePart) part;
        byte[] bytes = imagePart.imageBytes();
        if (bytes == null && imagePart.getFilePath() != null) {
          bytes = ImageDecoder.getImageBytes(context, imagePart.getFilePath());
        }
        if (bytes == null) {
          throw new IllegalArgumentException(
              "Failed to load image bytes from ImagePart: " + imagePart.getFilePath());
        }
        resolvedContent.add(bytes);
      } else if (part instanceof AudioPart) {
        AudioPart audioPart = (AudioPart) part;
        if (audioPart.audioData() != null) {
          resolvedContent.add(audioPart.audioData());
        } else if (audioPart.getFilePath() != null) {
          AudioData audio = AudioDecoder.decodeAudio(context, audioPart.getFilePath());
          if (audio == null) {
            throw new IllegalArgumentException(
                "Failed to load audio from URI: " + audioPart.getFilePath());
          }
          resolvedContent.add(audio);
        } else {
          throw new IllegalArgumentException("AudioPart must contain audioData or filePath.");
        }
      } else if (part instanceof TextPart) {
        TextPart textPart = (TextPart) part;
        resolvedContent.add(textPart.getText());
      } else {
        resolvedContent.add(part);
      }
    }

    for (EmbeddingProvider provider : components.providers()) {
      float[] embedding = provider.embedContent(resolvedContent);
      if (embedding != null) {
        return embedding;
      }
    }
    throw new IllegalStateException("No embedding provider found that can embed this content.");
  }

  /**
   * Retrieves the topK records from the vector store for the given query text.
   *
   * @param queryText the query text to retrieve records for.
   * @param topK the number of records to retrieve.
   * @return the topK records from the vector store for the given query text.
   */
  public List<RetrievalResult> retrieve(String queryText, int topK) {
    return retrieve(queryText, topK, /* metadataFilter= */ null);
  }

  /**
   * Retrieves the topK records from the vector store for the given query text and metadata filter.
   *
   * @param queryText the query text to retrieve records for.
   * @param topK the number of records to retrieve.
   * @param metadataFilter key-value pairs that candidate records must match.
   * @return the topK records from the vector store.
   */
  public List<RetrievalResult> retrieve(
      String queryText, int topK, Map<String, String> metadataFilter) {
    return retrieve(ImmutableList.of(new TextPart(queryText)), topK, metadataFilter);
  }

  /**
   * Retrieves the topK records from the vector store for the given query parts.
   *
   * @param queryParts the query parts to retrieve records for.
   * @param topK the number of records to retrieve.
   * @return the topK records from the vector store for the given query parts.
   */
  public List<RetrievalResult> retrieve(List<Part> queryParts, int topK) {
    return retrieve(queryParts, topK, /* metadataFilter= */ null);
  }

  /**
   * Retrieves the topK records from the vector store for the given query parts and metadata filter.
   *
   * @param queryParts the query parts to retrieve records for.
   * @param topK the number of records to retrieve.
   * @param metadataFilter key-value pairs that candidate records must match.
   * @return the topK records from the vector store.
   */
  public List<RetrievalResult> retrieve(
      List<Part> queryParts, int topK, Map<String, String> metadataFilter) {
    long timestamp = syntheticTimestamp.getAndIncrement();
    statsLogger.recordCpuInputArrival(timestamp);
    float[] queryEmbedding = embedContent(queryParts);
    List<RetrievalRecord> searchResults =
        components.vectorStore().search(queryEmbedding, topK, metadataFilter);

    List<RetrievalRecord> results = new ArrayList<>();
    Set<String> uniqueParentIds = new LinkedHashSet<>();
    for (RetrievalRecord record : searchResults) {
      String parentId = record.getParentId();
      if (parentId != null && !parentId.isEmpty()) {
        uniqueParentIds.add(parentId);
      }
    }

    Map<String, RetrievalRecord> parentRecordsMap = new HashMap<>();
    if (!uniqueParentIds.isEmpty()) {
      List<RetrievalRecord> parentRecords =
          components.vectorStore().get(ImmutableList.copyOf(uniqueParentIds));
      for (RetrievalRecord parent : parentRecords) {
        parentRecordsMap.put(parent.getId(), parent);
      }
    }

    Set<String> seenParents = new HashSet<>();
    for (RetrievalRecord record : searchResults) {
      String parentId = record.getParentId();
      if (parentId != null && !parentId.isEmpty()) {
        if (seenParents.add(parentId)) {
          RetrievalRecord parent = parentRecordsMap.get(parentId);
          if (parent != null) {
            // Populate the parent record with the highest-scoring child record's embedding.
            RetrievalRecord parentWithChildEmbedding =
                new RetrievalRecord(
                    parent.getId(),
                    parent.getContent(),
                    record.getEmbeddings(),
                    parent.getMetadata(),
                    parent.getParentId(),
                    parent.getChildIds());
            results.add(parentWithChildEmbedding);
          }
        }
      } else {
        if (seenParents.add(record.getId())) {
          results.add(record);
        }
      }
    }

    List<RetrievalResult> clientResults = new ArrayList<>();
    for (RetrievalRecord record : results) {
      Embedding query = Embedding.create(queryEmbedding, new byte[0], 0, Optional.empty());
      Embedding stored = Embedding.create(record.getEmbeddings(), new byte[0], 0, Optional.empty());
      double score = CosineSimilarity.compute(query, stored);
      clientResults.add(
          new RetrievalResult(record.getId(), record.getContent(), record.getMetadata(), score));
    }
    statsLogger.recordInvocationEnd(timestamp);
    return clientResults;
  }

  /**
   * Deletes the records from the vector store for the given ids.
   *
   * @param ids the ids of the records to delete.
   */
  public void delete(List<String> ids) {
    components.vectorStore().delete(ids);
  }

  /**
   * Deletes the records from the vector store matching the metadata filter.
   *
   * @param metadataFilter key-value pairs that records to delete must match.
   */
  public void delete(Map<String, String> metadataFilter) {
    components.vectorStore().delete(metadataFilter);
  }

  /** Deletes all records from the vector store. */
  public void deleteAll() {
    components.vectorStore().deleteAll();
  }

  /**
   * Retrieves all unique record identifiers in the vector store.
   *
   * @return a list of all unique record identifiers.
   */
  public List<String> getAllRecordIds() {
    return components.vectorStore().getAllRecordIds();
  }

  /** Closes the vector store. */
  @Override
  public void close() {
    statsLogger.logSessionEnd();
    if (components.vectorStore() != null) {
      try {
        components.vectorStore().close();
      } catch (Exception e) {
        // Ignore closing errors
      }
    }
  }
}

File: mediapipe/tasks/web/retrieval/semantic_retriever/semantic_retriever.ts
/**
 * Copyright 2026 The MediaPipe Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {computeCosineSimilarity} from '../../../../tasks/web/components/utils/cosine_similarity';
import type {
  ContentPart,
  EmbeddingProvider,
} from '../../../../tasks/web/core/embedding_provider';

import type {RetrievalOptions} from './retrieval_options';
import type {RetrievalRecord} from './retrieval_record';
import {SemanticRetrieverComponents} from './semantic_retriever_components';
import type {RetrievalResult} from './semantic_retriever_result';
import type {TextChunker} from './text_chunker';
import type {VectorStore} from './vector_store';

export {DefaultTextChunker} from './default_text_chunker';
export {MemoryVectorStore} from './memory_vector_store';
export type {RetrievalOptions} from './retrieval_options';
export type {RetrievalRecord} from './retrieval_record';
export {SemanticRetrieverComponents} from './semantic_retriever_components';
export type {RetrievalResult} from './semantic_retriever_result';
export type {ChunkingMode, TextChunker} from './text_chunker';
export type {VectorStore} from './vector_store';
export type {ContentPart, EmbeddingProvider};

/**
 * SemanticRetriever coordinates document chunking, multimodal embedding,
 * and high-dimensional vector search.
 */
export class SemanticRetriever {
  private isClosed = false;

  /**
   * Creates a SemanticRetriever from SemanticRetrieverComponents.
   *
   * @param components The SemanticRetriever components.
   * @return A Promise resolving to the SemanticRetriever instance.
   * @export
   */
  static async createFromComponents(
    components: SemanticRetrieverComponents,
  ): Promise<SemanticRetriever> {
    if (!components || components.providers().length === 0) {
      throw new Error(
        'At least one EmbeddingProvider must be registered in SemanticRetrieverComponents.',
      );
    }

    const vectorStore = components.vectorStore();
    if (!vectorStore) {
      throw new Error(
        'VectorStore must be provided in SemanticRetrieverComponents.',
      );
    }

    const textChunker = components.textChunker();
    if (!textChunker) {
      throw new Error(
        'TextChunker must be provided in SemanticRetrieverComponents.',
      );
    }

    return new SemanticRetriever(components, vectorStore, textChunker);
  }

  /** @hideconstructor */
  private constructor(
    private readonly components: SemanticRetrieverComponents,
    private readonly vectorStore: VectorStore,
    private readonly textChunker: TextChunker,
  ) {}

  private async embedContent(
    parts: readonly ContentPart[],
  ): Promise<Float32Array> {
    for (const provider of this.components.providers()) {
      const result = await provider.embedContent(parts);
      if (result != null) {
        return result instanceof Float32Array
          ? result
          : new Float32Array(result);
      }
    }
    throw new Error('No embedding provider found that can embed this content.');
  }

  /**
   * Inserts a text document into the retriever with automatic chunking.
   * @export
   */
  async insertDocument(
    id: string,
    text: string,
    metadata?: Record<string, string>,
  ): Promise<void> {
    this.ensureNotClosed();
    return this.insertContent(id, [{text}], metadata);
  }

  /**
   * Inserts multimodal content parts under a single unique ID.
   * @export
   */
  async insertContent(
    id: string,
    parts: readonly ContentPart[],
    metadata?: Record<string, string>,
  ): Promise<void> {
    this.ensureNotClosed();
    await this.vectorStore.delete([id]);

    const databaseRecords: Array<{
      id: string;
      parts: readonly ContentPart[];
      parentId?: string;
      childIds?: string[];
    }> = [];
    let childCount = 0;

    const isSingleTextPart =
      parts.length === 1 &&
      'text' in parts[0] &&
      typeof parts[0].text === 'string';

    if (isSingleTextPart) {
      const textPart = parts[0] as {text: string};
      const chunks = await this.textChunker.chunk(textPart.text);
      if (chunks.length > 1) {
        const childIds: string[] = [];
        for (const chunkText of chunks) {
          const childId = `${id}_chunk_${childCount++}`;
          childIds.push(childId);
          databaseRecords.push({
            id: childId,
            parts: [{text: chunkText}],
            parentId: id,
          });
        }
        databaseRecords.push({
          id,
          parts,
          childIds,
        });
      } else {
        databaseRecords.push({id, parts});
      }
    } else {
      databaseRecords.push({id, parts});
    }

    const recordsToUpsert: RetrievalRecord[] = [];
    for (const record of databaseRecords) {
      const isParentWithChildren =
        record.childIds !== undefined && record.childIds.length > 0;
      const embedding = isParentWithChildren
        ? new Float32Array(0)
        : await this.embedContent(record.parts);
      recordsToUpsert.push({
        id: record.id,
        content: [...record.parts],
        embeddings: embedding,
        metadata: metadata || {},
        parentId: record.parentId,
        childIds: record.childIds,
      });
    }
    await this.vectorStore.upsert(recordsToUpsert);
  }

  /**
   * Inserts an image into the retriever.
   * @export
   */
  async insertImage(
    id: string,
    imageBytes: Uint8Array,
    metadata?: Record<string, string>,
  ): Promise<void> {
    this.ensureNotClosed();
    return this.insertContent(id, [{imageBytes}], metadata);
  }

  /**
   * Inserts pre-decoded PCM float audio samples into the retriever.
   * @export
   */
  async insertAudio(
    id: string,
    audioData: Float32Array,
    metadata?: Record<string, string>,
  ): Promise<void> {
    this.ensureNotClosed();
    return this.insertContent(id, [{audioData}], metadata);
  }

  /**
   * Retrieves top-K nearest matching records based on a text or multimodal query.
   * @export
   */
  async retrieve(
    query: string | readonly ContentPart[],
    options: RetrievalOptions = {},
  ): Promise<RetrievalResult[]> {
    this.ensureNotClosed();
    const {limit = 5, metadataFilter} = options;
    if (limit <= 0) {
      throw new Error(`'limit' must be greater than 0, got ${limit}.`);
    }
    const queryParts: readonly ContentPart[] =
      typeof query === 'string' ? [{text: query}] : query;
    const queryEmbedding = await this.embedContent(queryParts);
    // Oversample candidates to avoid under-fetching during parent deduplication.
    const queryLimit = limit * 3;
    const searchResults = await this.vectorStore.search(
      queryEmbedding,
      queryLimit,
      metadataFilter,
    );

    const uniqueParentIds = new Set<string>();
    for (const record of searchResults) {
      if (record.parentId) {
        uniqueParentIds.add(record.parentId);
      }
    }

    const parentRecordsMap = new Map<string, RetrievalRecord>();
    if (uniqueParentIds.size > 0) {
      const parentRecords = await this.vectorStore.get(
        Array.from(uniqueParentIds),
      );
      for (const parent of parentRecords) {
        parentRecordsMap.set(parent.id, parent);
      }
    }

    const deduplicatedResults: RetrievalRecord[] = [];
    const seenParents = new Set<string>();
    for (const record of searchResults) {
      if (record.parentId) {
        if (!seenParents.has(record.parentId)) {
          seenParents.add(record.parentId);
          const parent = parentRecordsMap.get(record.parentId);
          if (parent) {
            deduplicatedResults.push({
              id: parent.id,
              content: parent.content,
              embeddings: record.embeddings,
              metadata: parent.metadata,
              parentId: parent.parentId,
              childIds: parent.childIds,
            });
          }
        }
      } else {
        if (!seenParents.has(record.id)) {
          seenParents.add(record.id);
          deduplicatedResults.push(record);
        }
      }
    }

    const clientResults: RetrievalResult[] = [];
    const queryEmbeddingArray = Array.from(queryEmbedding);
    for (const record of deduplicatedResults) {
      let score = 0;
      if (record.embeddings && record.embeddings.length > 0) {
        try {
          score = computeCosineSimilarity(
            {
              floatEmbedding: queryEmbeddingArray,
              headIndex: 0,
              headName: '',
            },
            {
              floatEmbedding: Array.from(record.embeddings),
              headIndex: 0,
              headName: '',
            },
          );
        } catch {
          score = 0;
        }
      }
      clientResults.push({
        id: record.id,
        content: record.content ? [...record.content] : [],
        metadata: record.metadata ? {...record.metadata} : {},
        score,
      });
      if (clientResults.length >= limit) {
        break;
      }
    }
    return clientResults;
  }

  /**
   * Deletes records and their chunks by ID(s), or records matching metadata filter.
   * @export
   */
  async delete(
    idsOrFilter: string | readonly string[] | Record<string, string>,
  ): Promise<void> {
    this.ensureNotClosed();
    if (typeof idsOrFilter === 'string') {
      await this.vectorStore.delete([idsOrFilter]);
    } else if (Array.isArray(idsOrFilter)) {
      await this.vectorStore.delete(idsOrFilter);
    } else {
      await this.vectorStore.deleteByMetadata(
        idsOrFilter as Record<string, string>,
      );
    }
  }

  /**
   * Retrieves all unique top-level record IDs stored in the retriever.
   * @export
   */
  async getAllRecordIds(): Promise<string[]> {
    this.ensureNotClosed();
    const ids = await this.vectorStore.getAllRecordIds();
    return Array.from(new Set(ids));
  }

  /**
   * Deletes all records from the retriever.
   * @export
   */
  async deleteAll(): Promise<void> {
    this.ensureNotClosed();
    await this.vectorStore.deleteAll();
  }

  /**
   * Closes the retriever and frees resources.
   * @export
   */
  close(): void {
    if (!this.isClosed) {
      this.isClosed = true;
      try {
        this.vectorStore.close();
      } catch {
        // Ignore closing errors.
      }
    }
  }

  private ensureNotClosed(): void {
    if (this.isClosed) {
      throw new Error('SemanticRetriever has already been closed.');
    }
  }
}


